ios/browser/api/bookmarks/exporter/bookmarks_encoder.cc

```diff
@@ -0,0 +1,191 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/ios/browser/api/bookmarks/exporter/bookmarks_encoder.h"
+#include "brave/ios/browser/api/bookmarks/exporter/exported_bookmark_entry.h"
+#include "components/bookmarks/browser/bookmark_codec.h"
+
+#include <stddef.h>
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/base64.h"
+#include "base/guid.h"
+#include "base/json/json_string_value_serializer.h"
+#include "base/memory/ptr_util.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/values.h"
+#include "components/bookmarks/browser/bookmark_model.h"
+#include "components/strings/grit/components_strings.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "url/gurl.h"
+
+using base::Time;
+
+using bookmarks::BookmarkCodec;
+
+namespace ios {
+// Current version of the file.
+static const int kCurrentVersion = 1;
+
+class BookmarkEncoder {
+public:
+  BookmarkEncoder();
+  ~BookmarkEncoder();
+  
+  std::unique_ptr<base::Value> Encode(
+      const ExportedBookmarkEntry* bookmark_bar_node,
+      const ExportedBookmarkEntry* other_folder_node,
+      const ExportedBookmarkEntry* mobile_folder_node,
+      const std::string& sync_metadata_str);
+  
+private:
+  std::unique_ptr<base::Value> EncodeNode(const ExportedBookmarkEntry* node);
+  
+  // Updates the check-sum with the given string.
+  void UpdateChecksum(const std::string& str);
+  void UpdateChecksum(const base::string16& str);
+
+  // Updates the check-sum with the given contents of URL/folder bookmark node.
+  // NOTE: These functions take in individual properties of a bookmark node
+  // instead of taking in a BookmarkNode for efficiency so that we don't convert
+  // various data-types to UTF16 strings multiple times - once for serializing
+  // and once for computing the check-sum.
+  // The url parameter should be a valid UTF8 string.
+  void UpdateChecksumWithUrlNode(const std::string& id,
+                                 const base::string16& title,
+                                 const std::string& url);
+  void UpdateChecksumWithFolderNode(const std::string& id,
+                                    const base::string16& title);
+
+  // Initializes/Finalizes the checksum.
+  void InitializeChecksum();
+  void FinalizeChecksum();
+  
+  // MD5 context used to compute MD5 hash of all bookmark data.
+  base::MD5Context md5_context_;
+  
+  // Checksums.
+  std::string computed_checksum_;
+  std::string stored_checksum_;
+};
+
+BookmarkEncoder::BookmarkEncoder() = default;
+
+BookmarkEncoder::~BookmarkEncoder() = default;
+
+void BookmarkEncoder::UpdateChecksum(const std::string& str) {
+  base::MD5Update(&md5_context_, str);
```

Comment:
md5 is not secure against an attacker who can modify the input to it. it should be ok if it's only used here to check if some data has changed, not for a security-related check.

Reply:
@diracdeltas as with the other code I commented on, this is copied from upstream. In this case we should actually be using the original BookmarkCodec class as I mentioned above